C7Engine/AI/BarbarianAI.cs

```diff
@@ -21,7 +21,7 @@ public class BarbarianAI {
 			// issue for the barbs but will be for similar loops elsewhere in the AI logic.
 			foreach (MapUnit unit in player.units.ToArray()) {
 				if (UnitIsFreeToMove(unit)) {
-					while (unit.movementPoints.canMove) {
+					while (unit.hitPointsRemaining > 0 && unit.movementPoints.canMove) {
```

Comment:
If this is the part that avoids the crash, I feel like it's just masking the problem. Why is the unit still showing up here once it's been killed? Is it specifically units that die before running out of MP, or are there already dead units still in player.units? Either way it seems like a more fundamental problem than is being addressed here.

Reply:
The unit is removed from player.units when it dies, but it is still in player.units.ToArray() from line 22.  The array is created because otherwise, we would have a ConcurrentModificationException (or the C# equivalent thereof) when removing it from player.units while iterating over that collection.

It is specifically units that die before they run out of MP.

I'm going to push up a change that results in the unit.move method returning false if the unit is killed while moving, allowing the while loop to be broken out of in that case.  It's more code, but means we won't keep iterating on the zombie unit for another `while`  (I'd also like to refactor some of the unit.move and animation mechanics at some point, but that might take past Halloween)